File: src/filter.js
'use strict';
const config = require('./config');
const { triggerN8n } = require('./n8nTrigger');

// ── Per-coin cooldown tracker ──────────────────────────────────────────────────
// Map<symbol, lastTriggerTimestamp>
const cooldownMap = new Map();

// ── Stats (exposed to /status) ────────────────────────────────────────────────
const stats = {
  totalChecked: 0,
  totalPassed: 0,
  totalBlocked: 0,
  perCoin: {},   // symbol → { triggered, lastAt, lastPercent }
};

/**
 * Called for every 24hrTicker event from Binance.
 * Applies filter rules and forwards to n8n if criteria met.
 *
 * Ticker fields we use:
 *   s  = symbol          (e.g. "BTCUSDT")
 *   P  = priceChangePercent (e.g. "4.23" or "-2.10")
 *   c  = lastPrice
 *   h  = highPrice
 *   l  = lowPrice
 *   v  = baseVolume
 *   q  = quoteVolume
 */
function applyFilter(ticker) {
  const symbol       = ticker.s;
  const pctChange    = parseFloat(ticker.P);
  const lastPrice    = parseFloat(ticker.c);
  const highPrice    = parseFloat(ticker.h);
  const lowPrice     = parseFloat(ticker.l);
  const baseVolume   = parseFloat(ticker.v);
  const quoteVolume  = parseFloat(ticker.q);

  stats.totalChecked++;

  // ── Rule 1: Price change threshold ────────────────────────────────────────
  if (Math.abs(pctChange) < config.PRICE_CHANGE_THRESHOLD) {
    return; // Filtered out
  }

  // ── Rule 2: Volume spike (if enabled) ─────────────────────────────────────
  if (config.VOLUME_SPIKE_MULTIPLIER > 0) {
    const baseline = getBaselineVolume(symbol);
    if (baseline !== null && quoteVolume < baseline * config.VOLUME_SPIKE_MULTIPLIER) {
      stats.totalBlocked++;
      return; // Volume didn't spike enough
    }
    // Update rolling baseline
    updateBaselineVolume(symbol, quoteVolume);
  }

  // ── Rule 3: Cooldown per coin ─────────────────────────────────────────────
  const lastTriggered = cooldownMap.get(symbol);
  const now = Date.now();
  if (lastTriggered && now - lastTriggered < config.COOLDOWN_MS) {
    return; // Still in cooldown
  }

  // ── PASSED all filters → trigger n8n ─────────────────────────────────────
  stats.totalPassed++;
  cooldownMap.set(symbol, now);

  if (!stats.perCoin[symbol]) {
    stats.perCoin[symbol] = { triggered: 0, lastAt: null, lastPercent: null };
  }
  stats.perCoin[symbol].triggered++;
  stats.perCoin[symbol].lastAt = new Date().toISOString();
  stats.perCoin[symbol].lastPercent = pctChange;

  const direction = pctChange >= 0 ? '🚀' : '📉';
  console.log(`[FILTER] ${direction} ${symbol} passed: ${pctChange >= 0 ? '+' : ''}${pctChange.toFixed(2)}% | Sending to n8n...`);

  triggerN8n({
    symbol,
    priceChangePercent: pctChange.toFixed(4),
    lastPrice: lastPrice.toFixed(8),
    highPrice: highPrice.toFixed(8),
    lowPrice: lowPrice.toFixed(8),
    volume: baseVolume.toFixed(4),
    quoteVolume: quoteVolume.toFixed(2),
    direction: pctChange >= 0 ? 'UP' : 'DOWN',
    triggeredAt: new Date().toISOString(),
  });
}

// ── Volume baseline (simple rolling average for spike detection) ───────────────
const volumeHistory = new Map(); // symbol → [..last 10 values]
function getBaselineVolume(symbol) {
  const history = volumeHistory.get(symbol);
  if (!history || history.length < 3) return null;
  return history.reduce((a, b) => a + b, 0) / history.length;
}
function updateBaselineVolume(symbol, val) {
  if (!volumeHistory.has(symbol)) volumeHistory.set(symbol, []);
  const h = volumeHistory.get(symbol);
  h.push(val);
  if (h.length > 10) h.shift(); // Keep last 10
}

function getStats() {
  return { ...stats };
}

module.exports = { applyFilter, getStats };

File: engine/filter.go
package engine

import (
	"log"
	"math"
	"strconv"
	"sync"
	"time"

	"express-websocket/config"
)

type TickerData struct {
	Symbol             string      `json:"s"`
	PriceChangePercent interface{} `json:"P"`
	LastPrice          interface{} `json:"c"`
	HighPrice          interface{} `json:"h"`
	LowPrice           interface{} `json:"l"`
	BaseVolume         interface{} `json:"v"`
	QuoteVolume        interface{} `json:"q"`
}

// Global Stats Map (Concurrency safe)
type StatsData struct {
	TotalChecked int64
	TotalPassed  int64
	TotalBlocked int64
	PerCoin      map[string]*CoinStat
	mu           sync.Mutex
}

type CoinStat struct {
	Triggered   int
	LastAt      string
	LastPercent float64
}

var Stats = StatsData{
	PerCoin: make(map[string]*CoinStat),
}

// Maps
var cooldownMap sync.Map
var volumeHistory sync.Map

func parseFloat(v interface{}) float64 {
	switch val := v.(type) {
	case float64:
		return val
	case string:
		f, _ := strconv.ParseFloat(val, 64)
		return f
	default:
		return 0
	}
}

func ApplyFilter(ticker TickerData) {
	pctChange := parseFloat(ticker.PriceChangePercent)
	lastPrice := parseFloat(ticker.LastPrice)
	highPrice := parseFloat(ticker.HighPrice)
	lowPrice := parseFloat(ticker.LowPrice)
	baseVolume := parseFloat(ticker.BaseVolume)
	quoteVolume := parseFloat(ticker.QuoteVolume)

	// Increment total checked
	Stats.mu.Lock()
	Stats.TotalChecked++
	Stats.mu.Unlock()

	// ── Rule 1: Price change threshold
	if math.Abs(pctChange) < config.AppConfig.PriceChangeThreshold {
		return // Filtered out
	}

	// ── Rule 2: Volume spike (if enabled)
	if config.AppConfig.VolumeSpikeMultiplier > 0 {
		baseline := getBaselineVolume(ticker.Symbol)
		if baseline > 0 && quoteVolume < baseline*config.AppConfig.VolumeSpikeMultiplier {
			Stats.mu.Lock()
			Stats.TotalBlocked++
			Stats.mu.Unlock()
			return // Volume didn't spike enough
		}
		// Update rolling baseline
		updateBaselineVolume(ticker.Symbol, quoteVolume)
	}

	// ── Rule 3: Cooldown per coin
	now := time.Now()
	if lastTriggeredRaw, ok := cooldownMap.Load(ticker.Symbol); ok {
		lastTriggered := lastTriggeredRaw.(time.Time)
		if now.Sub(lastTriggered).Milliseconds() < int64(config.AppConfig.CooldownMs) {
			return // Still in cooldown
		}
	}

	// ── PASSED all filters → trigger n8n
	cooldownMap.Store(ticker.Symbol, now)

	Stats.mu.Lock()
	Stats.TotalPassed++
	if _, exists := Stats.PerCoin[ticker.Symbol]; !exists {
		Stats.PerCoin[ticker.Symbol] = &CoinStat{}
	}
	Stats.PerCoin[ticker.Symbol].Triggered++
	Stats.PerCoin[ticker.Symbol].LastAt = now.Format(time.RFC3339)
	Stats.PerCoin[ticker.Symbol].LastPercent = pctChange
	Stats.mu.Unlock()

	direction := "📉"
	dirTag := "DOWN"
	sign := ""
	if pctChange >= 0 {
		direction = "🚀"
		dirTag = "UP"
		sign = "+"
	}

	log.Printf("[FILTER] %s %s passed: %s%.2f%% | Sending to n8n...\n",
		direction, ticker.Symbol, sign, pctChange)

	// Fire and forget n8n trigger
	go TriggerN8n(TriggerPayload{
		Symbol:             ticker.Symbol,
		PriceChangePercent: formatFloat(pctChange, 4),
		LastPrice:          formatFloat(lastPrice, 8),
		HighPrice:          formatFloat(highPrice, 8),
		LowPrice:           formatFloat(lowPrice, 8),
		Volume:             formatFloat(baseVolume, 4),
		QuoteVolume:        formatFloat(quoteVolume, 2),
		Direction:          dirTag,
		Threshold:          config.AppConfig.PriceChangeThreshold,
		TriggeredAt:        now.Format(time.RFC3339),
	})
}

func getBaselineVolume(symbol string) float64 {
	val, ok := volumeHistory.Load(symbol)
	if !ok {
		return 0
	}
	history := val.([]float64)
	if len(history) < 3 {
		return 0
	}
	sum := 0.0
	for _, v := range history {
		sum += v
	}
	return sum / float64(len(history))
}

func updateBaselineVolume(symbol string, val float64) {
	var history []float64
	stored, ok := volumeHistory.Load(symbol)
	if ok {
		history = stored.([]float64)
	}
	history = append(history, val)
	if len(history) > 10 {
		history = history[1:] // Keep last 10
	}
	volumeHistory.Store(symbol, history)
}

func formatFloat(v float64, prec int) string {
	return strconv.FormatFloat(v, 'f', prec, 64)
}
